src/src/main/java/org/niis/xroad/restadapterservice/ProviderGateway.java

```diff
@@ -251,6 +242,21 @@ protected ServiceResponse handleRequest(ServiceRequest request) throws SOAPExcep
 
     }
 
+    protected RESTClient getRESTClient(ProviderEndpoint endpoint) {
+        if (USE_PROXY) {
+            return RESTClientFactory.createRESTClient(endpoint.getHttpVerb(), HOVERFLY_PROXY_HOST, HOVERFLY_PROXY_PORT);
+        }
+        return RESTClientFactory.createRESTClient(endpoint.getHttpVerb());
+    }
+
+    protected void setProcessingWrappers(ProviderEndpoint endpoint, AbstractMessage... messages) {
```

Comment:
I think setProcessingWrappers(endpoint, request, response) would be better, since that's what it does. It it not used for managing 1...n AbstractMessages, and having method signature like that makes it's purpose less obvious for someone reading the code.

Reply:
Also, these changes reduce complexity, but I feel like the real problematic complexity in this class - which is more worth fixing - is elsewhere. Perhaps I would look at the logic starting from line ~207 `if (!RESTGatewayUtil.isValidContentType(contentType)) {`, ? And how about moving the internal classes ReqToMapRequestDeserializerImpl, DecryptingReqToMapRequestDeserializerImpl, XMLServiceResponseSerializer, EncryptingXMLServiceResponseSerializer to real classes, maybe under restadapterservice.util.providergateway....? Of course, if that is done, same should ideally be done for ConsumerGateway.

Reply:
Well I kinda like that the setting wrappers is generic as I did commit it even though generic code tends to be a bit more complex to read. It's true that the method is not used at the moment to modify more than two messages but what if in future? I would prefer to leave it as it is. Refactoring the whole class maybe out of scope as true as it is what @jansu76 pointed out regarding the complexity. Other oppinions?

Reply:
Certainly if we have some method that updates sometimes two items, sometimes more, it would make sense to make it more generic by using varags.

public Integer addNumbers(Integer...numbers)

However this should not be done always and I do not think that it should be done in this case.

Here we are setting request and response to process wrappers the same way as endpoint is.

```
        // Set request and response wrapper processing
        if (endpoint.isProcessingWrappers() != null) {
            request.setProcessingWrappers(endpoint.isProcessingWrappers());
            response.setProcessingWrappers(endpoint.isProcessingWrappers());
        }
```

In this case it's quite likely that only 2 things will ever be set: request and response. It is not likely that at some point in the future we will be setting 3 or 4 AbstractMessages. Using AbstractMessage instead of ServiceRequest and ServiceResponse, and using 1...n items instead of 2 makes the code (just a tiny bit) more abstract and more difficult to reason about, without giving any benefits if we accept that we will most likely always be working with just a request and response. This is an important, more general concept: one should not implement solutions that increase complexity or make the code harded to read and understand, if the requirement they are preparing for is hypothetical and not very likely. 

This is also important IMO since the PR is all about "reducing complexity". We should not increase the _real complexity_ even though we could a smaller complexity number out of SonarQube that way. If refactoring the whole class is unfeasible for this task, I would rather leave it for later and maybe not worry about the numbers SonarQube is giving. I don't think it is too valuable to focus on small changes here and there so that we pass some arbitrary complexity metric value that has happened to be selected for a criteria. SonarQube metrics are just a tool; the real goal is that the code is good quality and easy to read etc.

Reply:
Well yes, it's likely possible that there's no need in future either to set more than two messages so I'll change the method implementation.